arch/otbn/printer.s.dfy

```diff
@@ -292,37 +293,75 @@ method printCode(c: code, depth: int)
         case Ins32(ins) => printIndent(depth); printIns32(ins);
         case Ins256(ins) => printIndent(depth); printIns256(ins);
         case Block(block) => printBlock(block, depth);
+        case Function(name, fbody) =>
+            printIndent(depth); print("jal x1, "); print(name); print("\n");
         case While(wcond, wbody) =>
         {
             printIndent(depth); printWhileCond(wcond); print(", ");
             print(codeSize(wbody) + 1); print("\n"); // + 1 for nop instruction
             printCode(wbody, depth + 1);
-            printIndent(depth+1); print("  nop\n"); // ensures different end addrs for nested loops
+            printIndent(depth+1); print("nop\n"); // ensures different end addrs for nested loops
         }
         case Comment(com) => print(com);
 }
 
-method printProc(proc_name:seq<char>, code:code)
+method getFunctionsFromCodes(block: codes, defs: set<string>, res: seq<(string, codes)>) 
+    returns (defs': set<string>, res': seq<(string, codes)>) 
 {
-    print(".globl "); print(proc_name); print("\n");
-    print(proc_name); print(":\n");
-    printCode(code, 0);
-    print("  ret\n");
+    var i := block;
+    defs', res' := defs, res;
+
+    while (i.va_CCons?)
+        decreases i
+    {
+        defs', res' := getFunctions(i.hd, defs', res');
+        i := i.tl;
+    }
 }
 
-datatype AsmTarget = OTBN
-datatype PlatformTarget = Linux | MacOS
+method getFunctions(c: code, defs: set<string>, res: seq<(string, codes)>) 
+    returns (defs': set<string>, res': seq<(string, codes)>)
+{
+    defs', res' := defs, res;
+
+    match c 
+        case Block(block) =>
+            defs', res' := getFunctionsFromCodes(block, defs, res);
+        case Function(name, fbody) => {
+            if name !in defs {
+                defs' := defs + {name};
+                res' := res + [(name, fbody)];
+            }
+            defs', res' := getFunctionsFromCodes(fbody, defs', res');
+        }
+        case While(_, wbody) => 
+            defs', res' := getFunctions(wbody, defs', res');
+        case _ =>
+            defs', res' := defs, res;
+}
 
-function method procName(proc_name:seq<char>, suffix:seq<char>, asm:AsmTarget, platform:PlatformTarget):seq<char>
+method printProc(code:code)
+    requires code.Function?
 {
-    match platform
-        case Linux => proc_name
-        case MacOS => "_" + proc_name
+    print(".globl modexp_var"); print("\n");
+    var defs, res := getFunctions(code, {}, []); 
+    var i := 0;
+    while i < |res|
+    {
+        print(res[i].0); print(":\n");
+        printCode(Block(res[i].1), 1);
+        i := i + 1;
+        printIndent(1); print("ret\n\n");
+    }
 }
 
+datatype AsmTarget = OTBN
+datatype PlatformTarget = Linux | MacOS
+
 method PrintDemo(asm:AsmTarget, platform:PlatformTarget)
```

Comment:
I think we had talked about factoring this out into a separate file, so that the printer would be application-independent.

Reply:
Yeah I think the application independent part is related to the hard coding comment above. Maybe on the printer side, we always print a `dummy` target, which is inlined (ignored), then on the vale side, we can set the `dummy` to call a none-inlined function.

Reply:
Or we can have a selection table, but I think we have to hardcode the table content?

Reply:
Could we just pass the string name and `codes` in as parameters to the printer?